charms/worker/k8s/src/charm.py

```diff
@@ -372,41 +372,59 @@ def _get_valid_annotations(self) -> Optional[dict]:
 
         return annotations
 
-    def _configure_annotations(self, config: BootstrapConfig):
-        """Configure the annotations for the Canonical Kubernetes cluster.
+    def _get_cluster_config_from_charm_config(self) -> UserFacingClusterConfig:
+        """Retrieve the cluster config from charm configuration.
+
+        Returns:
+            UserFacingClusterConfig: The expected cluster configuration.
+        """
+        local_storage = LocalStorageConfig(
+            enabled=self.config.get("local_storage_enabled"),
+            local_path=self.config.get("local_storage_local_path"),
+            reclaim_policy=self.config.get("local_storage_reclaim_policy"),
+            set_default=self.config.get("local_storage_set_default"),
+        )
+
+        return UserFacingClusterConfig(
+            local_storage=local_storage,
+            annotations=self._get_valid_annotations(),
+        )
+
+    def _configure_cluster_config(self, config: BootstrapConfig):
+        """Configure the cluster config for the Canonical Kubernetes cluster.
 
         Args:
             config (BootstrapConfig): Configuration object to bootstrap the cluster.
         """
-        annotations = self._get_valid_annotations()
-        if annotations is None:
-            return
-
         if not config.cluster_config:
-            config.cluster_config = UserFacingClusterConfig(annotations=annotations)
+            config.cluster_config = self._get_cluster_config_from_charm_config()
         else:
-            config.cluster_config.annotations = annotations
+            charm_cluster_config = self._get_cluster_config_from_charm_config()
+            # Merge the charm and user cluster configurations
+            # The existing configuration takes precedence over the charm configuration as
+            # this information is retrieved elsewhere (e.g. cloud-provider).
+            merged_data = {**charm_cluster_config.dict(), **config.cluster_config.dict()}
+            config.cluster_config = UserFacingClusterConfig(**merged_data)
```

Comment:
I don't know if I'm super happy with this kind of merging of values as it somewhat is non-intuitive that the existing config takes precedence. I adapted here to the existing code.

Reply:
Tell me more here.  the `config.cluster_config` is what the snap currently understands
`charm_cluster_config` is the juju admin's view of the configuration. 

why does the snap config take precedence over the juju admin config?

Reply:
discussed offline - the code structure was not great and I updated it now to be more concise.